src/Component/src/Symfony/Security/OperationAccessChecker.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Resource\Symfony\Security;
+
+use Sylius\Resource\Context\Context;
+use Sylius\Resource\Metadata\Operation;
+use Sylius\Resource\Metadata\OperationAccessCheckerInterface;
+use Symfony\Component\ExpressionLanguage\ExpressionLanguage;
+use Symfony\Component\Security\Core\Authentication\AuthenticationTrustResolverInterface;
+use Symfony\Component\Security\Core\Authentication\Token\NullToken;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Core\Role\RoleHierarchyInterface;
+
+final class OperationAccessChecker implements OperationAccessCheckerInterface
+{
+    public function __construct(
+        private readonly ?ExpressionLanguage $expressionLanguage = null,
+        private readonly ?AuthenticationTrustResolverInterface $authenticationTrustResolver = null,
+        private readonly ?RoleHierarchyInterface $roleHierarchy = null,
+        private readonly ?TokenStorageInterface $tokenStorage = null,
+        private readonly ?AuthorizationCheckerInterface $authorizationChecker = null,
+    ) {
+    }
+
+    public function isGranted(Operation $operation, Context $context, array $extraVariables = []): bool
+    {
+        if (null === $this->tokenStorage || null === $this->authenticationTrustResolver) {
+            throw new \LogicException('The "symfony/security" library must be installed to use the "security" attribute.');
+        }
+
+        if (null === $this->expressionLanguage) {
+            throw new \LogicException('The "symfony/expression-language" library must be installed to use the "security" attribute.');
+        }
+
+        if (null === $expression = $operation->getSecurity()) {
+            return true;
+        }
+
+        $variables = array_merge($extraVariables, [
+            'trust_resolver' => $this->authenticationTrustResolver,
+            'auth_checker' => $this->authorizationChecker, // needed for the is_granted expression function
+        ]);
+
+        if (null === $token = $this->tokenStorage->getToken()) {
+            $token = new NullToken();
+        }
+
+        $variables = array_merge($variables, $this->getVariables($token));
```

Comment:
Why array over here
```php

        $variables = array_merge($extraVariables, [
            'trust_resolver' => $this->authenticationTrustResolver,
            'auth_checker' => $this->authorizationChecker, // needed for the is_granted expression function
        ]);
```

but method over here

```php
$this->getVariables($token)
```

```suggestion
        $variables = array_merge($variables, [
            'token' => $token,
            'user' => $token->getUser(),
            'roles' => $this->getEffectiveRoles($token),
        ]);
```

Reply:
1. I'm not following why `getEffectiveRoles` is split from SF example `getVariables()`. 
2. Simple `getRoles()` should be enough, adding `Effective` suggests implementation details on this method, which are un-important in this context.

Reply:
@diimpp this file is a copy of https://github.com/api-platform/core/blob/main/src/Symfony/Security/ResourceAccessChecker.php